src/mol-plugin-state/manager/focus-camera/focus-first-residue.ts

```diff
@@ -0,0 +1,162 @@
+            }
+            const averagePosition = Vec3.zero();
+            averagePosition[0] = sumX / AtomIndexs.length;
+            averagePosition[1] = sumY / AtomIndexs.length;
+            averagePosition[2] = sumZ / AtomIndexs.length;
+            return averagePosition;
+        } else {
+            return Vec3.create(caPositions[0], caPositions[1], caPositions[2]);
+        }
+    }
+
+}
+
+export function pcaFocus(plugin: PluginContext, options: Partial<CameraFocusOptions> & { principalAxes?: PrincipalAxes }, positionToFlip?: Vec3) {
+    if (options?.principalAxes) {
+        const { origin, dirA, dirB, dirC } = options.principalAxes.boxAxes;
+        let toFlip: String[] = [];
+        if (positionToFlip) {
+            toFlip = getAxesToFlip(positionToFlip, origin, dirA, dirB);
+        }
+        toFlip.forEach((axis)=>{
+            if (axis === 'aroundY') {
+                Vec3.negate(dirC, dirC);
+            } else if (axis === 'aroundX') {
+                Vec3.negate(dirA, dirA);
+                Vec3.negate(dirC, dirC);
+            }
+        });
+        if (plugin.canvas3d) {
+            plugin.canvas3d.camera.setState(Camera.createDefaultSnapshot());
+            const position = Vec3();
+            Vec3.scaleAndAdd(position, position, origin, 100);
+            plugin.canvas3d.camera.setState({ position }, 0);
+            const deltaDistance = Vec3();
+            Vec3.negate(deltaDistance, position);
+            if (Vec3.dot(deltaDistance, dirC) <= 0) {
+                Vec3.negate(plugin.canvas3d.camera.position, position);
+            }
+            const up = Vec3.create(0, 1, 0);
+            if (Vec3.dot(up, dirA) <= 0) {
+                Vec3.negate(plugin.canvas3d?.camera.up, plugin.canvas3d.camera.up);
+            }
+        }
+        const newPrincipalAxes = {
+            origin: origin,
+            dirA: dirA,
+            dirB: dirB,
+            dirC: dirC
+        };
+        return newPrincipalAxes;
+    }
+    return {
+        origin: Vec3.zero(),
+        dirA: Vec3.zero(),
+        dirB: Vec3.zero(),
+        dirC: Vec3.zero()
+    };
+}
+
+export function getPcaTransform(group: StructureComponentRef[]): [{principalAxes: PrincipalAxes}|undefined, Vec3|undefined] {
+    let isPolymer = true;
+    let positions = new Float32Array();
+    let positionToFlip = Vec3.zero();
+
+    if (group[0].cell.obj) {
+        const polymerStructure = group[0].cell.obj.data;
+        positions = getPolymerPositions(polymerStructure);
+        positionToFlip = getFirstResidueOrAveragePosition(polymerStructure, positions);
+        if (polymerStructure.units[0].props.polymerElements?.length === 0) isPolymer = false;
+    }
+
+    if (isPolymer) {
+        return [
+            { principalAxes: PrincipalAxes.ofPositions(positions) },
+            positionToFlip
+        ];
+    } else {
+        return [
+            undefined,
+            undefined
+        ];
+
+    }
```

Comment:
It would be good to cache the PCA transform once it has been computed. You can do this for example like this:

```ts
if ('_pcaTransformData' in polymerStructure.currentPropertyData) {
  return polymerStructure.currentPropertyData._pcaTransformData;
}

...

if (polymerStructure.units[0]?.props.polymerElements?.length) {
    polymerStructure.currentPropertyData._pcaTransformData = undefined;
    return undefined;
}

polymerStructure.currentPropertyData._pcaTransformData = { ... };
```

Reply:
For the caching, I have commented out for now. I find that the caching will cause the orientation to flip back and forth in some case, such as 1tqn. Even though the cache is successfully run and the data stored in it remain unchanged. When I try to read the data, one vector of axis changed for the opposite direction. I will be glad if you can help me find out what is happening here. Thanks in advance.

Reply:
Problem solved. Deepclone is used here to avoid data mutation.